### PR TITLE
chore (All), fix(typings)

### PR DESCRIPTION
**Pull request by gerbendekker**
This still contains bug, doing pull request to coop.
Please advise to check the src file changes because they contain assumption to correct strict typescript issues.

### CODE REVIEW

File: dist/amd/index-internal.d.ts [hunk 1]
```diff
@@ -1,11 +1,11 @@
...
+import { OpenIdConnectAuthorizeStep } from './open-id-connect-authorize-step';
+import { OpenIdConnectConfigurationManager } from './open-id-connect-configuration-manager';
+import { OpenIdConnectFactory } from './open-id-connect-factory';
+import { OpenIdConnectLogger } from './open-id-connect-logger';
+import { OpenIdConnectNavigationStrategies } from './open-id-connect-navigation-strategies';
+import { OpenIdConnectRouting } from './open-id-connect-routing';
+import { OpenIdConnectUserBlock } from './open-id-connect-user-block';
+import { OpenIdConnectUserDebug } from './open-id-connect-user-debug';
+export * from './internal-types';
+export { RedirectHandler, OpenIdConnectAuthorizeStep, OpenIdConnectConfigurationManager, OpenIdConnectFactory, OpenIdConnectLogger, OpenIdConnectNavigationStrategies, OpenIdConnectRouting, OpenIdConnectUserBlock, OpenIdConnectUserDebug };
```
Comment (shaunluttin) [inline dist/amd/index-internal.d.ts]: Could you please remove the `dist` directory from the pull request. That will make it easier to review.

File: karma.conf.js [hunk 2]
```diff
@@ -6,7 +6,7 @@ module.exports = function (config) {
         files: [
             { pattern: 'node_modules/babel-polyfill/browser.js' },
             { pattern: "src/**/*.ts" },
-            { pattern: "spec/**/*.spec.ts" }
+            { pattern: "test/**/*.spec.ts" }
```
Comment (shaunluttin) [inline karma.conf.js]: Why the change from `spec` to `test`? Is this a convention you prefer, or is there a behavior difference too?

File: package.json [hunk 3]
```diff
@@ -1,39 +1,45 @@
...
+  "private": true,
+  "bugs": {
+    "url": "https://github.com/shaunluttin/aurelia-open-id-connect/issues"
+  },
+  "license": "MIT",
   "author": "Shaun Luttin <shaun@bigfont.ca>",
-  "main": "dist/amd/index.js",
-  "types": "dist/amd/index.d.ts",
+  "main": "dist/commonjs/index.js",
+  "typings": "dist/commonjs/index.d.ts",
```
Comment (shaunluttin) [inline package.json]: Why the change from `types` to `typings`?

As far as I can tell from the documentation, `types` and `typings` are synonymous. https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html

> Note that the "typings" field is synonymous with "types", and could be used as well.

File: package.json [hunk 4]
```diff
@@ -1,39 +1,45 @@
...
+    "build:system": "cross-env tsc --project tsconfig.json --outDir dist/system --module system",
+    "postbuild:system": "cross-env copyfiles --up 1 src/**/*.html src/**/*.css dist/system",
+    "prebuild": "cross-env rimraf dist",
+    "build": "concurrently \"npm run build:amd\" \"npm run build:commonjs\" \"npm run build:es2015\" \"npm run build:native-modules\" \"npm run build:system\"",
+    "changelog": "cross-env conventional-changelog -p aurelia-open-id-connect -i doc/CHANGELOG.md -s",
+    "bump-version": "npm --no-git-tag-version version",
+    "preprepare-release": "cross-env npm run test",
+    "prepare-release": "cross-env npm run changelog && npm run build"
   },
-  "license": "MIT",
```
Comment (shaunluttin) [inline package.json]: Are you not a fan of the MIT license?

File: package.json [hunk 5]
```diff
@@ -42,23 +48,58 @@
...
-    "karma-phantomjs-launcher": "^1.0.4",
-    "karma-typescript": "^3.0.8",
-    "mocha": "^4.0.1",
-    "phantomjs": "^2.1.7",
-    "proxyquire": "^1.8.0",
-    "sinon": "^4.1.2",
-    "tslint": "^5.8.0",
-    "tslint-microsoft-contrib": "^5.0.1",
+    "@types/jasmine": "^2.5.53",
+    "@types/node": "^7.0.11",
```
Comment (shaunluttin) [inline package.json]: Note-to-self: Test this. Does using `Jasmine` instead of `Mocha/Chai/Sinon` just work?

File: src/index-internal.ts [hunk 6]
```diff
@@ -1,18 +1,18 @@
...
-import IredirectHandler from "./iredirect-handler";
-import OpenIdConnectAuthorizeStep from "./open-id-connect-authorize-step";
-import OpenIdConnectConfigurationManager from "./open-id-connect-configuration-manager";
-import OpenIdConnectFactory from "./open-id-connect-factory";
-import OpenIdConnectLogger from "./open-id-connect-logger";
-import OpenIdConnectNavigationStrategies from "./open-id-connect-navigation-strategies";
-import OpenIdConnectRouting from "./open-id-connect-routing";
-import OpenIdConnectUserBlock from "./open-id-connect-user-block";
-import OpenIdConnectUserDebug from "./open-id-connect-user-debug";
+import { RedirectHandler } from './redirect-handler';
```
Comment (shaunluttin) [inline src/index-internal.ts]: I think this change reflects the canonical linting rules/style that Aurelia uses for TypeScript. Is that right?

File: src/open-id-connect-configuration-manager.ts [hunk 7]
```diff
@@ -47,11 +47,11 @@ export default class {
         return this._userManagerSettings;
     }
 
-    public get redirectUri(): string {
+    public get redirectUri(): string | undefined {
```
Comment (shaunluttin) [inline src/open-id-connect-configuration-manager.ts]: Hmm. This is an interesting one for me. Why was the union with `undefined` necessary?

File: src/open-id-connect-navigation-strategies.ts [hunk 8]
```diff
@@ -42,7 +42,7 @@ export default class OpenIdConnectNavigationStrategies {
             // And we will receive one of two errors:
             // 'No matching state found in storage' or
             // 'No state in response'
-            const url: string = null;
+            const url: string | undefined = undefined;
```
Comment (shaunluttin) [inline src/open-id-connect-navigation-strategies.ts]: Note-to-self: Consider changing this to `this.userManager.signinSilentCallback(undefined)` instead of having an explanatory variable.

File: src/open-id-connect-navigation-value-converter.ts [hunk 9]
```diff
@@ -18,6 +18,8 @@ export class OpenIdConnectNavigationValueConverter {
       if (roles.indexOf(OpenIdConnectRoles.Authenticated) >= 0) {
         return user !== null;
       }
+
+      return false;
```
Comment (shaunluttin) [inline src/open-id-connect-navigation-value-converter.ts]: Hmm. Nice catch.

File: src/open-id-connect-routing.ts [hunk 10]
```diff
@@ -21,30 +18,30 @@ export default class OpenIdConnectRouting {
...
+            name: 'logInRedirectCallback',
             navigationStrategy: (instruction: NavigationInstruction) => {
                 if (this.isSilentLogin()) {
                     return this.openIdConnectNavigationStrategies.silentSignInCallback(instruction);
                 } else {
                     return this.openIdConnectNavigationStrategies.signInRedirectCallback(instruction);
                 }
             },
-            route: this.getPath(this.openIdConnectConfiguration.redirectUri),
+            route: this.getPath(this.openIdConnectConfiguration.redirectUri!),
```
Comment (shaunluttin) [inline src/open-id-connect-routing.ts]: Note-to-self: Think about the use of the bang operator here, because it might be that we can set useful defaults instead of `undefined`, thereby obviating the need for the bang operator.

File: typings.json [hunk 11]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (shaunluttin) [inline typings.json]: Do we need this, given that TypeScript now supports the @types organization?

File: tslint.json [hunk 12]
```diff
@@ -1,15 +1,18 @@
...
+    "no-shadowed-variable": false,
+    "no-string-literal": false,
+    "trailing-comma": [false],
+    "member-ordering": [true, {"order": "fields-first"}],
+    "array-type": [true, "array-simple"],
+    "arrow-parens": false,
+    "max-classes-per-file": [false],
+    "prefer-for-of": false
+  }
+}
```
Comment (shaunluttin) [inline tslint.json]: What's the source of this set of tslint rules? For instance, is it the canonical ruleset for Aurelia TypeScript projects?

File: tsconfig.test.json [hunk 13]
```diff
@@ -1,11 +0,0 @@
-{
```
Comment (shaunluttin) [inline tsconfig.test.json]: Note-to-self: Do the tests still work?

File: tsconfig.json [hunk 14]
```diff
@@ -1,26 +1,29 @@
...
+    "experimentalDecorators": true,
+    "noEmitHelpers": false
+  },
+  "exclude": [
+    ".vscode",
+    "dist",
+    "doc",
+    "test",
+    "node_modules"
+  ]
```
Comment (shaunluttin) [inline tsconfig.json]: Note-to-self: Consider these tsconfig changes. What is the intent? Do the tests work?

File: karma.conf.js (hunk 2 above)
Comment (gerbendekker) [reply karma.conf.js]: Works the same, it is used by aurelia repos.
My own thoughts about it, the folder test contains them and *.spec.ts is a way to describe them.

File: package.json [hunk 15]
```diff
@@ -1,39 +1,45 @@
...
+    "postbuild:native-modules": "cross-env copyfiles --up 1 src/**/*.html src/**/*.css dist/native-modules",
+    "prebuild:system": "cross-env rimraf dist/system",
+    "build:system": "cross-env tsc --project tsconfig.json --outDir dist/system --module system",
+    "postbuild:system": "cross-env copyfiles --up 1 src/**/*.html src/**/*.css dist/system",
+    "prebuild": "cross-env rimraf dist",
+    "build": "concurrently \"npm run build:amd\" \"npm run build:commonjs\" \"npm run build:es2015\" \"npm run build:native-modules\" \"npm run build:system\"",
+    "changelog": "cross-env conventional-changelog -p aurelia-open-id-connect -i doc/CHANGELOG.md -s",
+    "bump-version": "npm --no-git-tag-version version",
+    "preprepare-release": "cross-env npm run test",
+    "prepare-release": "cross-env npm run changelog && npm run build"
```
[the remark this answers is not in the document]
Comment (gerbendekker) [reply package.json]: Yep -> https://github.com/aurelia/ux

File: package.json (hunk 4 above)
Comment (gerbendekker) [reply package.json]: did not intended this

File: package.json (hunk 5 above)
Comment (gerbendekker) [reply package.json]: The test are not tested 😄

File: tsconfig.json (hunk 14 above)
Comment (shaunluttin) [reply tsconfig.json]: How do the tests work now?

File: src/index-internal.ts (hunk 6 above)
Comment (gerbendekker) [reply src/index-internal.ts]: Yes since this its carrying 'aurelia' in its name I applied the same coding conventions.
In typescript the difference between a interface and a class is smaller then for example in .NET.
Prefixing interfaces with an I is less common typescript.

File: src/open-id-connect-configuration-manager.ts (hunk 7 above)
Comment (gerbendekker) [reply src/open-id-connect-configuration-manager.ts]: I am using the strict mode of typescript, that warns about possible problems.
the ```get redirectUri()``` returns a type ```uri?: string;```.
Necessary? no depend on your 'tsconfig.json' settings.
I did changes in more places that need to be evaluated. like the bang operator.
```this.openIdConnectConfiguration.redirectUri``` TS warns for the possibility of undefined.
```this.openIdConnectConfiguration.redirectUri!``` is a way of saying I have seen/checked it and accept it.

File: src/open-id-connect-navigation-value-converter.ts (hunk 9 above)
Comment (gerbendekker) [reply src/open-id-connect-navigation-value-converter.ts]: Wasn't me, credits go to TS compiler.

File: typings.json (hunk 11 above)
Comment (gerbendekker) [reply typings.json]: No, please not! my bad.

File: tslint.json (hunk 12 above)
Comment (gerbendekker) [reply tslint.json]: [aurelia-ux](https://github.com/aurelia/ux/blob/master/tslint.json)

File: dist/amd/index-internal.d.ts (hunk 1 above)
Comment (gerbendekker) [reply dist/amd/index-internal.d.ts]: will do, tomorrow

File: package.json (hunk 3 above)
Comment (gerbendekker) [reply package.json]: You are right, what i did is take the files from https://github.com/aurelia/ux as the package base.
Thats the only reason.

File: src/open-id-connect-configuration-manager.ts (hunk 7 above)
Comment (shaunluttin) [reply src/open-id-connect-configuration-manager.ts]: Thank you for letting me know. Increasing the stringency of the TypeScript compiler will improve the code.

File: karma.conf.js (hunk 2 above)
Comment (shaunluttin) [reply karma.conf.js]: Sure. I am going to push to master the changes that you suggested. This might lead to some conflicts. Let me know if that is a problem. For instance, I will use `test` instead of the `spec` directory, and will change double quotes to single quotes.

File: package.json (hunk 4 above)
Comment (gerbendekker) [reply package.json]: Looks like I moved the MIT license to the top of this file.
It is still present.

File: typings.json (hunk 11 above)
Comment (gerbendekker) [reply typings.json]: The typings.json helps consumers in some scenarios to get the typings of the aurelia-open-id-connect package. 
For the rest, use @types instead.

**Comment by gerbendekker** [comment]
@shaunluttin I'm ready with the pull request.
The tests are working and successful.
The package is checked with 'aurelia-app-identity-server-aspnetcore' and working.

File: package.json (hunk 5 above)
Comment (gerbendekker) [reply package.json]: tests are working now

**Comment by gerbendekker** [comment]
1 item remains, I excluded the test folder from tsconfig.json.
If I remove it then it triggers a "is not under 'rootDir'" maybe still use the tsconfig.test.json but I leave that up to you.

**Comment by gerbendekker** [comment]
btw, [fix (DI): autoinject to inject - c0f56d4](https://github.com/shaunluttin/aurelia-open-id-connect/pull/44/commits/c0f56d42108fdce5d2a78340eb8f07581bbf09a7) could also be resolved with `"emitDecoratorMetadata": true` in 'tsconfig.json'.

> TypeScript's emitDecoratorMetadata flag causes the TypeScript compiler to **polyfill** the Metadata Reflection API and add a special decorator definition to the transpiled TypeScript code.

Because of the polyfill not always used in libs/plugins, but I leave that decision up to you to make.

**Comment by shaunluttin** [comment]
Thank you for this PR. 

I ended up cherry-picking many of the changes instead of merging the entire PR. You can see the changes in the 0.24.1 release. I have also updated the contributors list.

If there are other changes that you made that you think would be worth including, and that I did not cherry-pick, please either 

1. create a new PR or
2. fix the branch conflicts in this PR.

Thank you again. The ASP.NET Core project now builds without errors! 👏 👏 👏

**Comment by gerbendekker** [comment]
Your welcome, thanks for appreciating the work.
Its hard to make no branch conflicts when reorganizing a lot.
If all set, then it would be good to look at #8 .
